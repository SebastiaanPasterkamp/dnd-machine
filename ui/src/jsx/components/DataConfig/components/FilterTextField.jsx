import React from 'react';
import PropTypes from 'prop-types';

import ControlGroup from '../../ControlGroup';
import FieldSet from '../../FieldSet';
import InputField from '../../InputField';
import { ListComponent } from '../../ListComponent';
import TagContainer from '../../TagContainer';

export class FilterTextField extends React.Component
{
    constructor(props) {
        super(props);
        this.onFieldChange = this.onFieldChange.bind(this);
        this.onFilterChange = this.onFilterChange.bind(this);
    }

    onFieldChange(field) {
        const { setState } = this.props;
        setState({ field });
    }

    onFilterChange(filter) {
        const { setState } = this.props;
        setState({ filter });
    }

    render() {
        const { field, filter } = this.props;

        return (
            <FieldSet label="Filter textfield">
                <ControlGroup label="Field">
                    <InputField
                        placeholder="Field..."
                        value={field}
                        type="text"
                        setState={this.onFieldChange}
                    />
                </ControlGroup>
                <FieldSet label="Values">
                    <ListComponent
                        list={filter}
                        component={InputField}
                        newItem="auto"
                        setState={this.onFilterChange}
                    />
                </FieldSet>
            </FieldSet>
        );
    }
};

FilterTextField.propTypes = {
    field: PropTypes.string,
    filter: PropTypes.arrayOf(PropTypes.string),
};

FilterTextField.defaultProps = {
    field: '',
    filter: [],
};

export default FilterTextField;
