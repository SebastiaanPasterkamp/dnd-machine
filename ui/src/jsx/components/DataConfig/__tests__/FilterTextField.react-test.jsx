import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import FilterTextField from '../components/FilterTextField';

describe('Component: FilterTextField', () => {
    const fullProps = {
        type: 'textfield',
        field: 'attrib',
        options: ['foo', 'bar']
    };

    describe('when rendering', () => {
        const setState = jest.fn();

        it('should work with minimum props', () => {
            const tree = renderer.create(
                <FilterTextField
                    setState={setState}
                />
            ).toJSON();

            expect(tree).toMatchSnapshot();
        })

        it('should work with full props', () => {
            const tree = renderer.create(
                <FilterTextField
                    {...fullProps}
                    setState={setState}
                />
            ).toJSON();

            expect(tree).toMatchSnapshot();
        })
    });

    describe('when editing', () => {
        it('should emit changes in the field', () => {
            const setState = jest.fn();

            const wrapper = mount(
                <FilterTextField
                    setState={setState}
                />
            );

            wrapper.find('input[type="text"]').at(0).simulate('change', {
                target: { value: fullProps.field }
            });

            expect(setState).toBeCalledWith({
                type: fullProps.type,
                field: fullProps.field,
            });
        });

        it('should emit changes in the values', () => {
            const setState = jest.fn();

            const wrapper = mount(
                <FilterTextField
                    {...fullProps}
                    setState={setState}
                />
            );

            wrapper.find('input').at(1).simulate('change', {
                target: { value: 'FOO' }
            });

            expect(setState).toBeCalledWith({
                type: fullProps.type,
                options: ['FOO', 'bar']
            });
        });

        it('should add new values', () => {
            const setState = jest.fn();

            const wrapper = mount(
                <FilterTextField
                    {...fullProps}
                    setState={setState}
                />
            );

            wrapper.find('input').at(3).simulate('change', {
                target: { value: 'ruh' }
            });

            expect(setState).toBeCalledWith({
                type: fullProps.type,
                options: ['foo', 'bar', 'ruh']
            });
        });

        it('should remove existing values', () => {
            const setState = jest.fn();

            const wrapper = mount(
                <FilterTextField
                    {...fullProps}
                    setState={setState}
                />
            );

            wrapper.find('button[name="del"]').at(0).simulate('click');

            expect(setState).toBeCalledWith({
                type: fullProps.type,
                options: ['bar']
            });
        });
    });
});
