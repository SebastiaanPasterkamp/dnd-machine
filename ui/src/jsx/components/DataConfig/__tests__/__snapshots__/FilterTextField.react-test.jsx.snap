// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: FilterTextField when rendering should work with full props 1`] = `
<fieldset
  className="nice-form-group"
>
  <legend>
    Filter textfield
  </legend>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Field
    </span>
    <input
      className="nice-form-control"
      disabled={undefined}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Field..."
      type="text"
      value="attrib"
    />
  </div>
  <fieldset
    className="nice-form-group"
  >
    <legend>
      Values
    </legend>
    <ul
      className="list-component"
    >
      <li>
        <input
          className="nice-form-control"
          disabled={undefined}
          onChange={[Function]}
          onKeyPress={null}
          placeholder={undefined}
          type={undefined}
          value="foo"
        />
        <button
          className="nice-btn list-component__delete warning icon fa-minus"
          disabled={false}
          name="del"
          onClick={[Function]}
        >
          ​
        </button>
      </li>
      <li>
        <input
          className="nice-form-control"
          disabled={undefined}
          onChange={[Function]}
          onKeyPress={null}
          placeholder={undefined}
          type={undefined}
          value="bar"
        />
        <button
          className="nice-btn list-component__delete warning icon fa-minus"
          disabled={false}
          name="del"
          onClick={[Function]}
        >
          ​
        </button>
      </li>
      <li>
        <input
          className="nice-form-control"
          disabled={undefined}
          onChange={[Function]}
          onKeyPress={null}
          placeholder={undefined}
          type={undefined}
          value=""
        />
        <button
          className="nice-btn list-component__delete warning icon fa-minus"
          disabled={true}
          name="del"
          onClick={[Function]}
        >
          ​
        </button>
      </li>
    </ul>
  </fieldset>
</fieldset>
`;

exports[`Component: FilterTextField when rendering should work with minimum props 1`] = `
<fieldset
  className="nice-form-group"
>
  <legend>
    Filter textfield
  </legend>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Field
    </span>
    <input
      className="nice-form-control"
      disabled={undefined}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Field..."
      type="text"
      value=""
    />
  </div>
  <fieldset
    className="nice-form-group"
  >
    <legend>
      Values
    </legend>
    <ul
      className="list-component"
    >
      <li>
        <input
          className="nice-form-control"
          disabled={undefined}
          onChange={[Function]}
          onKeyPress={null}
          placeholder={undefined}
          type={undefined}
          value=""
        />
        <button
          className="nice-btn list-component__delete warning icon fa-minus"
          disabled={true}
          name="del"
          onClick={[Function]}
        >
          ​
        </button>
      </li>
    </ul>
  </fieldset>
</fieldset>
`;
